### ISSUE TITLE
Create a manifest of files copied by the 'install' target.

### ISSUE THREAD

**Issue by mikew-lunarg** [KhronosGroup/Vulkan-ValidationLayers#743]
The list of files copied from this repo using the CMake "install" target needs to be more clearly defined, and with specified locations that can be relied upon.  

Create a manifest (i.e.: filesystem specification) that defines which files should go where.

Example questions to be answered: ( ref [layers/CMakeLists.txt](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/layers/CMakeLists.txt#L60-L84) )
* The current implementation does not specify a "vulkan" subdirectory.
* Should all files from LAYER_UTIL_FILES be copied to an include directory?
* Which files are needed by custom layer developers (as opposed to validation layer-specific)?
* Of non-public validation layer files, where should they go?

For developers of custom layers: it would be helpful to understand what kind of layer you're developing, and exactly which files you need access to.

**Comment by jzulauf-lunarg** [comment]
Added bullet. 

Are there currently "validation layer-specific" files (in terms of guard string, namespace etc. such as the vk_typemap_helper.h) that would be of use to custom layer developers with some polish.  This list should be considered carefully as these would become effectively part of an API with ongoing support/consistency considerations.

**Comment by karl-lunarg** [comment]
I have some thoughts to consider:
### How We Got Here
The current configuration is probably a result of the recent repo splits/reorgs.  The mechanism for a repo to get its dependencies is by using the install image generated by the repos it depends on.

The LunarG/VulkanTools repo is the source for several non-validation layers that need this set of layer include files.  This set of layer include files originate in Vulkan-ValidationLayers.  I think we discussed just replicating this set of layer include files in VulkanTools, but that idea was rejected for probably good reasons.

So instead, the V-VL repo was changed to "publish"/install these headers to "anywhere" in the V-VL install image and VT started depending on V-VL for these headers.  In the context of making these files available to repos that depend on this one, it didn't really matter where in the install include filesystem hierarchy these ended up in.  But it did have the (then) unintended side effect of Linux packaging publish-installing these files to `CMAKE_INSTALL_PREFIX/include`, which is `/usr/local/include` by default.  This isn't the right place and is the basis of the problem represented by #740 and #729.

The process for "developing a layer" outside of the repos was defined to use the SDK to get access to the "(validation) layer framework" which provided these layer include files.  I don't think we ever formally offered a way to develop layers or provide these include files in any other manner.

### Proposed Filesystem Location for Layer Header Files
There is some precedent in the X11 and GL include file structures for introducing subdirectories for containing specialized include files.  X11 has things like `X11/extensions` and `X11/Xft`.  GL has `GL/internal`.   I propose `<CMAKE_INSTALL_PREFIX>/include/vulkan/layers` as the location for these files.   "These files" means files that are needed to compile layers and are not validation-layer-specific.   Validation-layer-specific headers should stay in V-VL.  We may have to inform or confer with the Vulkan Working Group to see if this selection of `vulkan/layers` needs to be codified in the spec.

### First Solution Step
> The list of files copied from this repo using the CMake "install" target needs to be more clearly defined

The current CMake code is already authoritative here.  But if it helps to add CMake code to put these files in a CMake variable like `list_of_include_files_to_install`, then we can do that.  It probably makes the list of files look less accidental.  Of course, the list of files needs to be reviewed for correctness.

> and with specified locations that can be relied upon.

Again, explicit in the CMake code.  If it helps to add comments and/or words to BUILD.md that says "This repo installs layer header files to `<CMAKE_INSTALL_PREFIX>/include/vulkan/layers`" then we should do that.

I think that the 4 bullets have been or will be addressed here.

>For developers of custom layers: it would be helpful to understand what kind of layer you're developing, and exactly which files you need access to.

Hopefully, this should be driven by the owner/designer of the layer mechanism in this repo.  IOW, we should be taking at least the first shot at providing the right headers.  The VulkanTools repo is a good "customer" example that can be used for getting this right. 

Finally, a first pass at fixing this would be to:
1) Finallize the choice for location of headers (WG approval if needed)
2) Make sure the list of files is correct
3) Change this repo to install the files to `<CMAKE_INSTALL_PREFIX>/include/vulkan/layers`
4) Change VulkanTools to get the include files from the new location (known-good allows decoupling)

Sadly, this doesn't address other repos that may rely on the current location, but we don't know if there are any.  Need to update docs to highlight the change.

### "Final" Solution
Might be able to combine the previous "First" solution with this one if we want to do all this at once.

It probably makes sense to move the set of layer include files to the Vulkan-Headers repo.  If any of them are code-generated from the Vulkan Registry, then those generation scripts would move there too and some build-time targets added to Vulkan-Headers to build these before the Vulkan-Headers install target can be built.

Doing this move has some nice advantages:
* It seems more intuitive to put these headers in Vulkan-Headers, especially after formalizing their location
* Since these headers are not specific to the validation layers, they shouldn't come from the V-VL repo
* The V-VL repo no longer needs to house/generate these headers
* Repos like VulkanTools that depend on V-VL only for these headers would no longer need to depend on V-VL, which can lead to great savings in terms of building prerequisites.
* From a Linux packaging perspective, these headers can start appearing in existing packages (like "vulkan-devel") without creating new packages like "vulkan-layer-devel".  This is admittedly due more to the new location
* From a Linux packaging perspective, packages don't have to dig the layer header files out of V-VL.  They get them from Vulkan-Headers, which should be much nicer.
* Possible de-duplication of codegen scripts, etc

There are some potential downsides.  The layer include files become more exposed and are perhaps less flexible to changes.  Additional testing is likely necessary to make sure that the layer headers can be used, etc.

This whole effort formalizes the layer include files as being part of the overall Vulkan interface, as part of recognizing that Vulkan application developers are going to be developing layers as well as applications.  The layer include files therefore need additional attention.

It may also make sense to create a new CMake target in Vulkan-Headers named something like "layer-headers" which could be configured as INTERFACE PRIVATE to represent that these headers are needed to *build* the layers, but are not needed by applications to *use* the layers.  Applications use the layers by telling the loader to load them and they don't use header files to access/link the layers.  This target would also be useful to the "nested" build proposals.

**Comment by 1ace** [comment]
All of this is good, and should be done, but let's remember the original issue and try to solve it as well: there is (currently, and even with everything mentioned so far in this issue done) no way to *know* where those headers are. The best one could do is poke around in a few locations and hope one of those was used, and assume the headers don't exist otherwise.

In the GNU/Linux world, this is done by providing a [`pkg-config`](https://www.freedesktop.org/wiki/Software/pkg-config/) file. Once such a file is provided, it almost doesn't matter anymore where the headers are installed (and that's part of the point).

**Comment by mikew-lunarg** [comment]
@1ace Yes, solving the original issue is definitely a goal!  A motivation for creating this new issue was the original appeared to be a point solution to a larger problem.

In addition to pkg-config, there perhaps may be some refactoring of files/filesystem/packaging, and documentation along the lines of the [Linux FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard)

**Comment by mikew-lunarg** [comment]
Other repos also have similar requirements for a manifest and defined filesystem structure, see above.
Could create an umbrella task to coordinate a cross-repo solution as appropriate.

**Comment by jzulauf-lunarg** [comment]
This also can have documentation issues to correlate text references with file locations e.g. https://vulkan.lunarg.com/issue/view/5c6b0a8ee46ffe085085610a

**Comment by mikew-lunarg** [comment]
possible related https://vulkan.lunarg.com/issue/view/5c7dbb46e46ffe7d52e37970

**Comment by mark-lunarg** [comment]
Any updates here?  What's the next step?

**Comment by mikew-lunarg** [comment]
It's on my todo list, to be done.

**Comment by mikew-lunarg** [comment]
Collecting links to other/related resources:
- https://github.com/KhronosGroup/Vulkan-Loader/pull/195
- https://www.freedesktop.org/wiki/Specifications/basedir-spec/
- https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard
